torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -164,6 +189,55 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
         self._setattr_on_modules(self.sharded_param)
         self.sharded_state = ShardedState.SHARDED
 
+    @torch.no_grad()
+    def init_all_gather_output(
+        self,
+        all_gather_input_numel: int,
+        world_size: int,
+        dtype: torch.dtype,
+        device: torch.device,
+    ):
+        if self.all_gather_output.numel() > 0:
+            return  # already initialized
+        all_gather_output_size = torch.Size([all_gather_input_numel * world_size])
+        self.all_gather_output = torch.empty(
+            all_gather_output_size, dtype=dtype, device=device
+        )
+
+    @torch.no_grad()
+    def init_unsharded_param(self):
+        if hasattr(self, "_unsharded_param"):
```

Comment:
I was not sure what the best way to do this 'deferred' attribute initialization. A few notes:
- For `self.all_gather_output`, I initialized it to a size-0 tensor because in the current implementation, we could call `FSDPParam.to_sharded()` before the 1st all-gather (from state dict), which would be a no-op but `to_sharded()` would expect `all_gather_output` to be defined. (In the newest RFC PR for fusing communication groups, I deferred registering the state dict hooks to lazy initialization, in which case this would no longer be a problem.)
- If we initialize either `self.all_gather_output` or `self._unsharded_param` to `None`, then the typing becomes more unwieldy as we need to `assert` that the attribute is not `None` before using it to satisfy mypy. This was the main reason why I preferred to not let it become `Optional[nn.Parameter]`.